pkg/image/external/external_test.go

```diff
@@ -0,0 +1,244 @@
+package external
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rancher/rke/types/kdm"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	k3s            = "k3s"
+	rancherVersion = "v2.6.4"
+	k3sWebVersion  = "v1.23.6+k3s1"
+	rke2WebVersion = "v1.23.6+rke2r1"
+	rke2           = "rke2"
+	rke1           = "rke1"
+	devKDM         = "https://github.com/rancher/kontainer-driver-metadata/raw/dev-v2.6/data/data.json"
+	releaseKDM     = "https://releases.rancher.com/kontainer-driver-metadata/release-v2.6/data.json"
+)
+
+func TestGetExternalImages(t *testing.T) {
+	kubeSemVer := &semver.Version{
+		Major: 1,
+		Minor: 21,
+		Patch: 0,
+	}
+
+	type args struct {
+		rancherVersion           string
+		externalData             map[string]interface{}
+		source                   Source
+		minimumKubernetesVersion *semver.Version
+		kdmUrl                   string
+		image1                   string
+		image2                   string
+		image3                   string
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		want    []string
+		wantErr bool
+	}{
+		{
+			name: "k3s-test",
+			args: args{
+				rancherVersion:           rancherVersion,
+				externalData:             map[string]interface{}{},
+				source:                   k3s,
+				minimumKubernetesVersion: kubeSemVer,
+				kdmUrl:                   devKDM,
+				image1:                   "rancher/klipper-lb:v0.3.5",
+				image2:                   "rancher/mirrored-pause:3.6",
+				image3:                   "rancher/mirrored-metrics-server:v0.5.2",
+			},
+			wantErr: false,
+		},
+		{
+			name: "rke2-test",
+			args: args{
+				rancherVersion:           rancherVersion,
+				externalData:             map[string]interface{}{},
+				source:                   rke2,
+				minimumKubernetesVersion: kubeSemVer,
+				kdmUrl:                   releaseKDM,
+				image1:                   "rancher/pause:3.6",
+				image2:                   "rancher/rke2-runtime:v1.23.6-rke2r1",
+				image3:                   "rancher/rke2-cloud-provider:v0.0.3-build20211118",
+			},
+			wantErr: false,
+		},
+		{
+			name: "rke1-test-fail",
+			args: args{
+				rancherVersion:           rancherVersion,
+				externalData:             map[string]interface{}{},
+				source:                   rke1,
+				minimumKubernetesVersion: kubeSemVer,
+				kdmUrl:                   releaseKDM,
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			a := assert.New(t)
+			get, err := http.Get(tt.args.kdmUrl)
```

Comment:
Would it be possible to test this functionality without making HTTP calls? From my experience with the chart's airgap exporter code, this approach can lead to flaky tests and delay Rancher builds. For this reason the chart's tests were rewritten using a unit test approach that does not make any calls to the outside (HTTP calls, or git clones, etc).

Reply:
For standalone unit tests, the http calls are required. In order to test the `GetExternalImages` func it requires a kdm Data struct from `rancher/rke/types/kdm` - https://github.com/rancher/rancher/blob/release/v2.6/pkg/image/export/main.go#L68-L80. I don't see a way to work around this without mocking the entire kdm Data struct, which I would argue is one thing we should not mock as that is a constantly changing data point for Rancher. 

If the http calls are failing that would more than likely mean that Github is down, which means our builds will fail entirely. These unit tests are the only way to ensure that we are loading all of the required image lists which is what bit us this release with RKE2 Windows images missing and causing all RKE2 Windows airgap testing to fail and many cycles spent trying to figure out why.